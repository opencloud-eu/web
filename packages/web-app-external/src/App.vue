<template>
  <div v-if="isLoading" class="oc-height-1-1 oc-width-1-1 oc-flex oc-flex-center oc-flex-middle">
    <oc-spinner size="large" />
  </div>
  <iframe
    v-else-if="appUrl && method === 'GET'"
    :src="appUrl"
    class="oc-width-1-1 oc-height-1-1"
    :title="iFrameTitle"
    allowfullscreen
  />
  <div v-if="appUrl && method === 'POST' && formParameters" class="oc-height-1-1 oc-width-1-1">
    <form :action="appUrl" target="app-iframe" method="post">
      <input ref="subm" type="submit" :value="formParameters" class="oc-hidden" />
      <div v-for="(item, key, index) in formParameters" :key="index">
        <input :name="key" :value="item" type="hidden" />
      </div>
    </form>
    <iframe
      name="app-iframe"
      class="oc-width-1-1 oc-height-1-1"
      :title="iFrameTitle"
      allowfullscreen
    />
  </div>
</template>

<script setup lang="ts">
import { stringify } from 'qs'
import { computed, unref, nextTick, ref, watch, onMounted, useTemplateRef } from 'vue'
import { useTask } from 'vue-concurrency'
import { useGettext } from 'vue3-gettext'
import isEmpty from 'lodash-es/isEmpty'
import {
  getPermissionsForSpaceMember,
  GraphSharePermission,
  Resource,
  SpaceResource,
  isProjectSpaceResource,
  isPublicSpaceResource,
  isShareSpaceResource
} from '@opencloud-eu/web-client'
import { urlJoin } from '@opencloud-eu/web-client'
import {
  isSameResource,
  useCapabilityStore,
  useConfigStore,
  useMessages,
  useRequest,
  useAppProviderService,
  useRoute,
  queryItemAsString,
  useRouteQuery,
  useUserStore,
  getSharedDriveItem,
  setCurrentUserShareSpacePermissions,
  useSpacesStore,
  useClientService,
  useSharesStore
} from '@opencloud-eu/web-pkg'

const { space, resource, isReadOnly } = defineProps<{
  space: SpaceResource
  resource: Resource
  isReadOnly: boolean
}>()

defineEmits(['save', 'close']) // these are inherited from the AppWrapper.vue

const language = useGettext()
const { $gettext } = language
const { showErrorMessage } = useMessages()
const capabilityStore = useCapabilityStore()
const configStore = useConfigStore()
const route = useRoute()
const appProviderService = useAppProviderService()
const { makeRequest } = useRequest()
const userStore = useUserStore()
const spacesStore = useSpacesStore()
const sharesStore = useSharesStore()
const { graphAuthenticated: graphClient } = useClientService()

const viewModeQuery = useRouteQuery('view_mode')
const viewModeQueryValue = computed(() => {
  return queryItemAsString(unref(viewModeQuery))
})

const templateIdQuery = useRouteQuery('templateId')
const templateIdQueryValue = computed(() => {
  return queryItemAsString(unref(templateIdQuery))
})

const appName = computed(() => {
  const lowerCaseAppName = unref(route)
    .name.toString()
    .replace('external-', '')
    .replace('-apps', '')
  return appProviderService.appNames.find((appName) => appName.toLowerCase() === lowerCaseAppName)
})

const appUrl = ref<string>()
const formParameters = ref({})
const method = ref<string>()
const subm = useTemplateRef<HTMLInputElement>('subm')
const isLoading = computed(() => loadAppUrl.isRunning || getSharedDriveItemTask.isRunning)

const iFrameTitle = computed(() => {
  return $gettext('"%{appName}" app content area', {
    appName: unref(appName)
  })
})

const errorPopup = (error: string) => {
  showErrorMessage({
    title: $gettext('An error occurred'),
    desc: error,
    errors: [new Error(error)]
  })
}

const loadAppUrl = useTask(function* (signal, viewMode: string) {
  try {
    if (isReadOnly && viewMode === 'write') {
      showErrorMessage({ title: $gettext('Cannot open file in edit mode as it is read-only') })
      return
    }

    const fileId = resource.fileId
    const baseUrl = urlJoin(configStore.serverUrl, capabilityStore.filesAppProviders[0].open_url)

    const query = stringify({
      file_id: fileId,
      lang: language.current,
      ...(unref(appName) && { app_name: encodeURIComponent(unref(appName)) }),
      ...(viewMode && { view_mode: viewMode }),
      ...(unref(templateIdQueryValue) && { template_id: unref(templateIdQueryValue) })
    })

    const url = `${baseUrl}?${query}`
    const response = yield makeRequest('POST', url, {
      validateStatus: () => true,
      signal
    })

    if (response.status !== 200) {
      switch (response.status) {
        case 425:
          errorPopup(
            $gettext(
              'This file is currently being processed and is not yet available for use. Please try again shortly.'
            )
          )
          break
        default:
          errorPopup(response.data?.message)
      }

      throw new Error('Error fetching app information')
    }

    if (!response.data.app_url || !response.data.method) {
      throw new Error('Error in app server response')
    }

    appUrl.value = response.data.app_url
    method.value = response.data.method

    if (response.data.form_parameters) {
      formParameters.value = response.data.form_parameters
    }

    if (method.value === 'POST' && formParameters.value) {
      yield nextTick()
      unref(subm).click()
    }
  } catch (e) {
    console.error('web-app-external error', e)
    throw e
  }
}).restartable()

const getSharedDriveItemTask = useTask(function* (signal) {
  try {
    return getSharedDriveItem({ graphClient, spacesStore, space, signal })
  } catch (e) {
    console.error(e)
  }
})

const determineOpenAsPreview = (appName: string) => {
  const openAsPreview = configStore.options.editor.openAsPreview
  return openAsPreview === true || (Array.isArray(openAsPreview) && openAsPreview.includes(appName))
}

// switch to write mode when edit is clicked
const catchClickMicrosoftEdit = (event: MessageEvent) => {
  try {
    if (JSON.parse(event.data)?.MessageId === 'UI_Edit') {
      loadAppUrl.perform('write')
    }
  } catch {}
}
onMounted(() => {
  if (determineOpenAsPreview(unref(appName))) {
    window.addEventListener('message', catchClickMicrosoftEdit)
  } else {
    window.removeEventListener('message', catchClickMicrosoftEdit)
  }
})

watch(
  [resource],
  async ([newResource], [oldResource]) => {
    if (isSameResource(newResource, oldResource)) {
      return
    }

    let viewMode = 'read'

    if (isShareSpaceResource(space)) {
      // add current user as space member if not already loaded by the space loader
      if (isEmpty(space.members)) {
        const sharedDriveItem = await getSharedDriveItemTask.perform()
        setCurrentUserShareSpacePermissions({
          sharesStore,
          spacesStore,
          userStore,
          space,
          sharedDriveItem
        })
      }

      const permissions = getPermissionsForSpaceMember(space, userStore.user)
      if (!permissions.includes(GraphSharePermission.readContent)) {
        // secure view
        viewMode = 'view'
      }
    }

    if (!isReadOnly) {
      viewMode = unref(viewModeQueryValue) || 'write'
    }

    if (
      determineOpenAsPreview(unref(appName)) &&
      (isShareSpaceResource(space) || isPublicSpaceResource(space) || isProjectSpaceResource(space))
    ) {
      viewMode = 'view'
    }
    loadAppUrl.perform(viewMode)
  },
  { immediate: true, deep: true }
)
</script>
