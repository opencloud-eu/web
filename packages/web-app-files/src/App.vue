<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="files-wrapper oc-width-expand oc-height-1-1 oc-flex-wrap" />
  </main>
</template>
<script lang="ts">
import { defineComponent, onBeforeUnmount, watch, ref } from 'vue'
import { useRoute, eventBus, useResourcesStore } from '@opencloud-eu/web-pkg'

export default defineComponent({
  setup() {
    const dragareaEnabled = ref(false)
    const { resetSelection } = useResourcesStore()

    watch(useRoute(), () => {
      resetSelection()
    })

    const hideDropzone = () => {
      dragareaEnabled.value = false
    }
    const onDragOver = (event: DragEvent) => {
      dragareaEnabled.value = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }

    const dragOver = eventBus.subscribe('drag-over', onDragOver)
    const dragOut = eventBus.subscribe('drag-out', hideDropzone)
    const drop = eventBus.subscribe('drop', hideDropzone)

    onBeforeUnmount(() => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
    return { dragareaEnabled }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-role-outline);
}

#files {
  position: relative;
}
</style>
