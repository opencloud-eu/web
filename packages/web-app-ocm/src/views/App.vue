<template>
  <div class="sciencemesh overflow-auto">
    <div class="flex flex-col h-full">
      <div class="grid grid-cols-1 md:grid-cols-2 h-auto max-h-auto md:max-h-[360px]">
        <div
          id="sciencemesh-invite"
          class="m-2 p-2 mb-0 lg:mb-2 bg-role-surface-container rounded-xl overflow-auto"
        >
          <outgoing-invitations />
        </div>
        <div
          id="sciencemesh-accept-invites"
          class="m-2 p-2 bg-role-surface-container rounded-xl overflow-auto"
        >
          <incoming-invitations @highlight-new-connections="highlightNewConnections" />
        </div>
      </div>
      <div id="sciencemesh-connections" class="p-2 bg-role-surface-container rounded-xl m-2 flex-1">
        <connections-panel
          v-model:connections="connections"
          :highlighted-connections="highlightedConnections.map((c) => c.id)"
          :loading="loadingConnections"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, onUnmounted, ref, unref, Ref } from 'vue'
import ConnectionsPanel from './ConnectionsPanel.vue'
import IncomingInvitations from './IncomingInvitations.vue'
import OutgoingInvitations from './OutgoingInvitations.vue'
import {
  useClientService,
  useScrollTo,
  FederatedConnection,
  useMessages
} from '@opencloud-eu/web-pkg'
import { useGettext } from 'vue3-gettext'
import { buildConnection } from '../functions'

export default defineComponent({
  components: {
    IncomingInvitations,
    OutgoingInvitations,
    ConnectionsPanel
  },
  setup() {
    const { showMessage } = useMessages()
    const { scrollToResource } = useScrollTo()
    const clientSerivce = useClientService()
    const { $gettext } = useGettext()

    const connections: Ref<FederatedConnection[]> = ref([])
    const highlightedConnections: Ref<FederatedConnection[]> = ref([])
    const highlightNewConnectionsInterval = ref<ReturnType<typeof setInterval>>()
    const loadingConnections = ref(true)

    const findAcceptedUsers = async () => {
      try {
        const { data: acceptedUsers } = await clientSerivce.httpAuthenticated.get<
          FederatedConnection[]
        >('/sciencemesh/find-accepted-users')
        loadingConnections.value = false
        connections.value = acceptedUsers.map(buildConnection)
      } catch {
        connections.value = []
        loadingConnections.value = false
      }
    }

    const highlightNewConnections = async () => {
      const oldConnections = [...unref(connections)]
      await findAcceptedUsers()
      if (oldConnections.length < unref(connections).length) {
        highlightedConnections.value = unref(connections).filter(
          (connection) => !oldConnections.map((c) => c.id).includes(connection.id)
        )
        if (unref(highlightedConnections).length === 1) {
          scrollToResource(unref(highlightedConnections)[0].id)
          showMessage({
            title: $gettext('New federated connection'),
            status: 'success',
            desc: $gettext('You can share with and recieve shares from %{user} now', {
              user: unref(highlightedConnections)[0].display_name
            })
          })
        } else if (unref(highlightedConnections).length > 1) {
          const newConnections = unref(highlightedConnections)
            .map((c) => c.display_name)
            .join(', ')

          showMessage({
            title: $gettext('New federated connections'),
            status: 'success',
            desc: $gettext('You can share with and receive shares from %{ connections } now', {
              connections: newConnections
            })
          })
        }
      }
    }

    onMounted(async () => {
      await findAcceptedUsers()
      loadingConnections.value = false
      highlightNewConnectionsInterval.value = setInterval(() => {
        highlightNewConnections()
      }, 10 * 1000)
    })

    onUnmounted(() => {
      clearInterval(unref(highlightNewConnectionsInterval))
    })

    return {
      highlightNewConnections,
      connections,
      highlightedConnections,
      loadingConnections
    }
  }
})
</script>
