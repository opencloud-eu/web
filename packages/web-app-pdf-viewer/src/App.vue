<template>
  <object class="pdf-viewer oc-width-1-1 oc-height-1-1" :data="url" :type="objectType" />
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  props: {
    url: {
      type: String,
      required: true
    }
  },
  setup() {
    const isSafari =
      navigator.userAgent?.includes('Safari') && !navigator.userAgent?.includes('Chrome')
    const objectType = isSafari ? undefined : 'application/pdf'

    return {
      objectType
    }
  }
})
</script>

<style scoped>
.pdf-viewer {
  overflow: hidden;
}
</style>
