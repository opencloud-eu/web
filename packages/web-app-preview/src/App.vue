<template>
  <div v-if="isFolderLoading" class="oc-width-1-1">
    <div class="oc-position-center">
      <oc-spinner :aria-label="$gettext('Loading media file')" size="xlarge" />
    </div>
  </div>
  <div
    v-else
    ref="preview"
    class="oc-flex oc-width-1-1 oc-height-1-1"
    tabindex="-1"
    @keydown.left="goToPrev"
    @keydown.right="goToNext"
  >
    <div class="stage" :class="{ lightbox: isFullScreenModeActivated }">
      <div class="stage_media">
        <div v-if="!activeMediaFileCached || activeMediaFileCached.isLoading" class="oc-width-1-1">
          <div class="oc-position-center">
            <oc-spinner :aria-label="$gettext('Loading media file')" size="xlarge" />
          </div>
        </div>
        <div
          v-else-if="activeMediaFileCached.isError"
          class="oc-width-1-1 oc-flex oc-flex-column oc-flex-middle oc-flex-center"
        >
          <oc-icon name="file-damage" size="xlarge" color="var(--oc-role-error)" />
          <p>
            {{ $gettext('Failed to load "%{filename}"', { filename: activeMediaFileCached.name }) }}
          </p>
        </div>
        <media-image
          v-else-if="activeMediaFileCached.isImage"
          :file="activeMediaFileCached"
          :current-image-rotation="currentImageRotation"
          :current-image-zoom="currentImageZoom"
          :current-image-position-x="currentImagePositionX"
          :current-image-position-y="currentImagePositionY"
          @pan-zoom-change="onPanZoomChanged"
        />
        <media-video
          v-else-if="activeMediaFileCached.isVideo"
          :file="activeMediaFileCached"
          :is-auto-play-enabled="isAutoPlayEnabled"
        />
        <media-audio
          v-else-if="activeMediaFileCached.isAudio"
          :file="activeMediaFileCached"
          :resource="activeFilteredFile"
          :is-auto-play-enabled="isAutoPlayEnabled"
        />
      </div>
      <media-controls
        class="stage_controls"
        :files="filteredFiles"
        :active-index="activeIndex"
        :is-full-screen-mode-activated="isFullScreenModeActivated"
        :is-folder-loading="isFolderLoading"
        :show-image-controls="activeMediaFileCached?.isImage && !activeMediaFileCached?.isError"
        :show-delete-button="isDeleteButtonVisible"
        :current-image-rotation="currentImageRotation"
        :current-image-zoom="currentImageZoom"
        @set-rotation="currentImageRotation = $event"
        @set-zoom="currentImageZoom = $event"
        @reset-image="resetImage"
        @toggle-full-screen="toggleFullScreenMode"
        @toggle-previous="goToPrev"
        @toggle-next="goToNext"
        @delete-resource="$emit('delete:resource')"
      />
    </div>
  </div>
</template>
<script lang="ts">
import {
  computed,
  defineComponent,
  ref,
  unref,
  PropType,
  nextTick,
  getCurrentInstance,
  watch,
  Ref
} from 'vue'
import omit from 'lodash-es/omit'
import { IncomingShareResource, Resource } from '@opencloud-eu/web-client'
import {
  AppFileHandlingResult,
  AppFolderHandlingResult,
  FileContext,
  ProcessorType,
  SortDir,
  createFileRouteOptions,
  queryItemAsString,
  sortHelper,
  useRoute,
  useRouteQuery,
  useRouter,
  usePreviewService,
  useGetMatchingSpace,
  isLocationSharesActive,
  useAppNavigation,
  useKeyboardActions,
  Modifier,
  Key,
  useFileActionsDelete
} from '@opencloud-eu/web-pkg'
import MediaControls from './components/MediaControls.vue'
import MediaAudio from './components/Sources/MediaAudio.vue'
import MediaImage from './components/Sources/MediaImage.vue'
import MediaVideo from './components/Sources/MediaVideo.vue'
import { CachedFile } from './helpers/types'
import {
  useFileTypes,
  useFullScreenMode,
  useImageControls,
  usePreviewDimensions
} from './composables'
import { mimeTypes } from './mimeTypes'

export const appId = 'preview'
const PRELOAD_COUNT = 5

export default defineComponent({
  name: 'Preview',
  components: {
    MediaControls,
    MediaAudio,
    MediaImage,
    MediaVideo
  },
  props: {
    activeFiles: { type: Object as PropType<Resource[]>, required: true },
    currentFileContext: { type: Object as PropType<FileContext>, required: true },
    loadFolderForFileContext: {
      type: Function as PropType<AppFolderHandlingResult['loadFolderForFileContext']>,
      required: true
    },
    getUrlForResource: {
      type: Function as PropType<AppFileHandlingResult['getUrlForResource']>,
      required: true
    },
    revokeUrl: { type: Function as PropType<AppFileHandlingResult['revokeUrl']>, required: true },
    isFolderLoading: { type: Boolean, required: true }
  },
  emits: ['update:resource', 'register:onDeleteResourceCallback', 'delete:resource'],
  setup(props, { emit }) {
    const router = useRouter()
    const route = useRoute()
    const contextRouteQuery = useRouteQuery('contextRouteQuery') as unknown as Ref<
      Record<string, string>
    >

    const { isFileTypeAudio, isFileTypeImage, isFileTypeVideo } = useFileTypes()
    const previewService = usePreviewService()
    const { dimensions } = usePreviewDimensions()
    const { getMatchingSpace } = useGetMatchingSpace()
    const { closeApp } = useAppNavigation({ router, currentFileContext: props.currentFileContext })
    const { bindKeyAction, removeKeyAction } = useKeyboardActions()
    const { actions: deleteFileActions } = useFileActionsDelete()

    const activeIndex = ref<number>()
    const cachedFiles = ref<Record<string, CachedFile>>({})
    const folderLoaded = ref(false)
    const isAutoPlayEnabled = ref(true)
    const preview = ref<HTMLElement>()
    const keyBindings: string[] = []

    const space = computed(() => {
      if (!unref(activeFilteredFile)) {
        return null
      }
      return getMatchingSpace(unref(activeFilteredFile))
    })

    const isDeleteButtonVisible = computed(() => {
      if (!unref(space)) {
        return false
      }
      return unref(deleteFileActions)[0]?.isVisible({
        space: unref(space),
        resources: [unref(activeFilteredFile)]
      })
    })

    const sortBy = computed(() => {
      if (!unref(contextRouteQuery)) {
        return 'name'
      }
      return unref(contextRouteQuery)['sort-by'] ?? 'name'
    })
    const sortDir = computed<SortDir>(() => {
      if (!unref(contextRouteQuery)) {
        return SortDir.Desc
      }
      return (unref(contextRouteQuery)['sort-dir'] as SortDir) ?? SortDir.Asc
    })

    const fileIdQuery = useRouteQuery('fileId')
    const fileId = computed(() => queryItemAsString(unref(fileIdQuery)))

    const filteredFiles = computed(() => {
      if (!props.activeFiles) {
        return []
      }

      const files = props.activeFiles.filter((file) => {
        if (
          unref(props.currentFileContext.routeQuery)['q_share-visibility'] === 'hidden' &&
          !(file as IncomingShareResource).hidden
        ) {
          return false
        }

        if (
          unref(props.currentFileContext.routeQuery)['q_share-visibility'] !== 'hidden' &&
          (file as IncomingShareResource).hidden
        ) {
          return false
        }

        return mimeTypes.includes(file.mimeType?.toLowerCase()) && file.canDownload()
      })

      return sortHelper(files, [{ name: unref(sortBy) }], unref(sortBy), unref(sortDir))
    })
    const activeFilteredFile = computed(() => {
      return unref(filteredFiles)[unref(activeIndex)]
    })
    const activeMediaFileCached = computed(() => {
      return unref(cachedFiles)[unref(activeFilteredFile)?.id]
    })

    const loadFileIntoCache = async (file: Resource) => {
      if (Object.hasOwn(unref(cachedFiles), file.id)) {
        return
      }

      const cachedFile: CachedFile = {
        id: file.id,
        name: file.name,
        url: undefined,
        ext: file.extension,
        mimeType: file.mimeType,
        isVideo: isFileTypeVideo(file),
        isImage: isFileTypeImage(file),
        isAudio: isFileTypeAudio(file),
        isLoading: ref(true),
        isError: ref(false)
      }
      cachedFiles.value[file.id] = cachedFile

      try {
        if (cachedFile.isImage) {
          cachedFile.url = await previewService.loadPreview(
            {
              space: unref(space),
              resource: file,
              dimensions: unref(dimensions),
              processor: ProcessorType.enum.fit
            },
            false,
            false
          )
          return
        }
        cachedFile.url = await props.getUrlForResource(unref(space), file)
      } catch (e) {
        console.error(e)
        cachedFile.isError.value = true
      } finally {
        cachedFile.isLoading.value = false
      }
    }

    const goToNext = () => {
      if (unref(activeIndex) + 1 >= unref(filteredFiles).length) {
        activeIndex.value = 0
        updateLocalHistory()
        return
      }
      activeIndex.value = unref(activeIndex) + 1
      updateLocalHistory()
    }

    const goToPrev = () => {
      if (unref(activeIndex) === 0) {
        activeIndex.value = unref(filteredFiles).length - 1
        updateLocalHistory()
        return
      }
      activeIndex.value = unref(activeIndex) - 1
      updateLocalHistory()
    }

    const onDeleteResourceCallback = async () => {
      await nextTick()

      if (!unref(filteredFiles).length) {
        return closeApp()
      }
    }

    const updateLocalHistory = () => {
      // this is a rare edge case when browsing quickly through a lot of files
      // we workaround context being null, when useDriveResolver is in loading state
      if (!props.currentFileContext) {
        return
      }

      const { params, query } = createFileRouteOptions(unref(space), unref(activeFilteredFile))
      router.replace({
        ...omit(unref(route), 'fullPath'),
        path: unref(route).fullPath,
        params: { ...unref(route).params, ...params },
        query: { ...unref(route).query, ...query }
      })
    }

    const instance = getCurrentInstance()

    watch(
      () => props.currentFileContext,
      async () => {
        if (!props.currentFileContext) {
          return
        }

        if (!unref(folderLoaded)) {
          await props.loadFolderForFileContext(props.currentFileContext)
          folderLoaded.value = true
        }

        ;(instance.proxy as any).setActiveFile()
      },
      { immediate: true }
    )

    watch(filteredFiles, () => {
      if (!unref(filteredFiles).length) {
        return
      }

      if (unref(activeIndex) >= unref(filteredFiles).length) {
        activeIndex.value = 0
        updateLocalHistory()
      }
    })

    watch(activeFilteredFile, (file) => {
      if (!file) {
        return
      }

      emit('update:resource', file)
    })

    const loading = computed(() => {
      if (props.isFolderLoading) {
        return true
      }
      const file = unref(activeMediaFileCached)
      if (!file) {
        return true
      }
      return unref(file.isLoading)
    })
    watch(
      loading,
      async (loading) => {
        if (!loading) {
          await nextTick()
          unref(preview).focus()
        }
      },
      { immediate: true }
    )

    return {
      ...useImageControls(),
      ...useFullScreenMode(),
      fileId,
      activeFilteredFile,
      activeIndex,
      activeMediaFileCached,
      cachedFiles,
      filteredFiles,
      updateLocalHistory,
      isAutoPlayEnabled,
      preview,
      isFileTypeImage,
      loadFileIntoCache,
      space,
      onDeleteResourceCallback,
      goToNext,
      goToPrev,
      keyBindings,
      bindKeyAction,
      removeKeyAction,
      isDeleteButtonVisible
    }
  },

  watch: {
    activeIndex(newValue, oldValue) {
      if (newValue !== oldValue) {
        this.loadFileIntoCache(this.activeFilteredFile)
        this.preloadImages()
      }

      if (oldValue !== null) {
        this.isAutoPlayEnabled = false
      }

      this.currentImageZoom = 1
      this.currentImageRotation = 0
    }
  },

  mounted() {
    // keep a local history for this component
    window.addEventListener('popstate', this.handleLocalHistoryEvent)
    this.$emit('register:onDeleteResourceCallback', this.onDeleteResourceCallback)
    this.keyBindings.push(
      this.bindKeyAction({ modifier: Modifier.Ctrl, primary: Key.Backspace }, () =>
        this.$emit('delete:resource', this.activeFilteredFile)
      )
    )
    this.keyBindings.push(
      this.bindKeyAction({ primary: Key.Delete }, () =>
        this.$emit('delete:resource', this.activeFilteredFile)
      )
    )
  },

  beforeUnmount() {
    window.removeEventListener('popstate', this.handleLocalHistoryEvent)
    this.keyBindings.forEach((keyBindingId) => {
      this.removeKeyAction(keyBindingId)
    })

    Object.values(this.cachedFiles).forEach((cachedFile) => {
      this.revokeUrl(cachedFile.url)
    })
  },

  methods: {
    setActiveFile() {
      for (let i = 0; i < this.filteredFiles.length; i++) {
        const filterAttr = isLocationSharesActive(this.$router, 'files-shares-with-me')
          ? 'remoteItemId'
          : 'fileId'

        // match the given file id with the filtered files to get the current index
        if (this.filteredFiles[i][filterAttr] === this.fileId) {
          this.activeIndex = i
          return
        }

        this.activeIndex = 0
      }
    },
    // react to PopStateEvent ()
    handleLocalHistoryEvent() {
      this.setActiveFile()
    },
    preloadImages() {
      const preloadFile = (preloadFileIndex: number) => {
        const cycleIndex =
          (((this.activeIndex + preloadFileIndex) % this.filteredFiles.length) +
            this.filteredFiles.length) %
          this.filteredFiles.length

        const file = this.filteredFiles[cycleIndex]
        this.loadFileIntoCache(file)
      }

      for (let followingFileIndex = 1; followingFileIndex <= PRELOAD_COUNT; followingFileIndex++) {
        preloadFile(followingFileIndex)
      }

      for (
        let previousFileIndex = -1;
        previousFileIndex >= PRELOAD_COUNT * -1;
        previousFileIndex--
      ) {
        preloadFile(previousFileIndex)
      }
    }
  }
})
</script>

<style lang="scss" scoped>
.stage {
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  text-align: center;

  &_media {
    flex-grow: 1;
    overflow: hidden;
    width: 100%;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &_controls {
    height: auto;
    margin: 10px auto;
  }
}
</style>
