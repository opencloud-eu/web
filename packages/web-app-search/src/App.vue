<template>
  <main id="search" class="flex h-full">
    <router-view id="search-view" />
  </main>
</template>
