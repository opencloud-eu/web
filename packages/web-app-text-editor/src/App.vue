<template>
  <div
    class="oc-text-editor oc-width-1-1 oc-height-1-1"
    :class="{ 'oc-text-editor-readonly': isReadOnly }"
  >
    <text-editor-component
      :resource="resource"
      :application-config="applicationConfig"
      :current-content="currentContent"
      :is-read-only="isReadOnly"
      @update:current-content="$emit('update:currentContent', $event)"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { Resource } from '@opencloud-eu/web-client'
import { AppConfigObject, TextEditor as TextEditorComponent } from '@opencloud-eu/web-pkg'

export default defineComponent({
  name: 'TextEditor',
  components: { TextEditorComponent },
  props: {
    applicationConfig: { type: Object as PropType<AppConfigObject>, required: true },
    currentContent: {
      type: String,
      required: true
    },
    isReadOnly: { type: Boolean, required: false },
    resource: { type: Object as PropType<Resource>, required: true }
  },
  emits: ['update:currentContent']
})
</script>

<style lang="scss">
.oc-text-editor-readonly {
  //Fixes in readonly mode vertical scrolling is not available
  height: calc(100vh - 52px);
  overflow: auto;
  padding: var(--oc-space-medium);
}
</style>
