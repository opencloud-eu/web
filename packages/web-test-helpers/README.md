# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
