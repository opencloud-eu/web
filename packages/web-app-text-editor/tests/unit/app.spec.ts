import { PartialComponentProps, mount } from '@opencloud-eu/web-test-helpers'
import App from '../../src/App.vue'

vi.mock('@opencloud-eu/web-pkg')

describe('Text editor app', () => {
  it('shows the editor', () => {
    const { wrapper } = getWrapper({
      applicationConfig: {}
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper(props: PartialComponentProps<typeof App>) {
  return {
    wrapper: mount(App, {
      props: {
        applicationConfig: {},
        currentContent: '',
        isReadOnly: false,
        resource: undefined,
        ...props
      }
    })
  }
}
