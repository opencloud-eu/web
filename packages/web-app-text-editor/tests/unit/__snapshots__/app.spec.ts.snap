// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text editor app > shows the editor 1`] = `
"<div class="oc-text-editor w-full oc-height-1-1">
  <!---->
</div>"
`;
